### DISCUSSION TITLE
How to cahe whole chapters / Create new Latex preset.

### DISCUSSION THREAD

**Discussion by dflvunoooooo** [leo-colisson/robust-externalize#37] [Q&A] [answered]
I am struggling to create my own latex preset to cache whole chapters. I want to create a latex preset with documentclass scrreport and set a few the docomentclass options. And with that preset cache an tex file which contains a chapter of latex. 

I tried it this way:
```
\documentclass{scrreport}

\usepackage{robust-externalize}

\robExtConfigure{
    new preset={latex scrreport}{
        set documentclass={scrreport},
        set latex options={fontsize=11pt a4paper},
    }
}

\begin{filecontents}{test2.tex}   
    \chapter{Test1}
    Test bla Test.   
\end{filecontents}

\begin{document}



\cacheMe[latex scrreport]{\input{../test2.tex}}


\end{document}
```

But I get the error `Package pgfkeys Error: I do not know the key '/robExt/set documentclass'` and the same for set latex options. Where did I go wrong?

**Comment by tobiasBora** [comment]
Sorry, I don't have access to a computer to give you a full working example, but a few remarks:
- you want to load `latex` before the `set documentclass`, as this option is created by the latex preset, hence your error
- I guess you need a comma between your options
- by default, we include stuff with `\includegraphics`, but for pdf pages it's a different command, something like `\includepdf[pages=-]{…}` from the `pdfpages` package, after loading the appropriate package, so you need to call `custom include command={\includepdf[pages=-]{\robExtAddCachePathAndName{\robExtFinalHash.pdf}}}`.

**Reply by dflvunoooooo** [reply]
Thank you for your answer. 
I see. I assumed that latex loads the standalone class and that will conflict with my choice of documentclass.
I will try it with one option until it is running. 
Thank you. I don't understand why this is needed, since a normal latex environment works as well. Is the normal cached latex inserted as an image? 

Now I get the error `! Improper \prevdepth.
\newpage ...everypar {}\fi \par \ifdim \prevdepth` . I don't know what this means. This happens with or without the custom inculde command.  

If I load the latex file inside a CacheMe environment, then I get the error `Paragraph ended before \@iinput was complete`.

```
\documentclass{scrreport}

\usepackage{robust-externalize}
\usepackage{pdfpages}

\robExtConfigure{
    new preset={latex scrreport}{
        latex, 
        set documentclass={scrreport},
        set latex options={fontsize=11pt},
%        custom include command={\includepdf{\robExtAddCachePathAndName{\robExtFinalHash.pdf}}}
    }
}

\begin{filecontents}{test2.tex}   
    \chapter{Test1}
    Test bla Test.   
\end{filecontents}

\begin{document}


\cacheMe[latex scrreport]{\input{__ROBEXT_WAY_BACK__test2.tex}}
%\begin{CacheMe}\input{__ROBEXT_WAY_BACK__test2.tex}\end{CacheMe}

\end{document}
```

**Reply by tobiasBora** [reply, marked as answer]
Ah good point, I forgot that by default I wrap the content with a box to measure its depth. Can you try adding `set placeholder={__ROBEXT_LATEX_MAIN_CONTENT_WRAPPED__}{__ROBEXT_MAIN_CONTENT__}`? Sorry I still can't test right now.

**Reply by tobiasBora** [reply]
> I assumed that latex loads the standalone class and that will conflict with my choice of documentclass

If you want to know exactly what `latex` is doing, you can check the code ^^' https://github.com/leo-colisson/robust-externalize/blob/2363dc1914e1816fc0850e2b5616cfbdcb9d4dec/robust-externalize.sty#L3482

But basically, when you create a preset from scratch, you always need to specify the compilation command, the template, and (often) the inclusion command. In your above code, you specified none of them. If you start it with `latex`, it will specify a template, a compilation command and an inclusion command, and to allow you to customize it, the template contains a number of placeholders, one of them containing the class. The above command therefore override the default standalone class to your own.

**Reply by dflvunoooooo** [reply]
This is working. Thank you very much! :) But, one has to use your custom include command. 

This is a working example:
```
\documentclass{scrreport}

\usepackage{robust-externalize}
\usepackage{pdfpages}

\robExtConfigure{
    new preset={latex scrreport}{
        latex, 
        set documentclass={scrreport},
        set latex options={fontsize=11pt, a4paper, version=last, english,french,ngerman, parskip=false, DIV=calc, captions=tableheading, pagesize, headinclude=true, twoside=false,},
        set placeholder={__ROBEXT_LATEX_MAIN_CONTENT_WRAPPED__}{__ROBEXT_MAIN_CONTENT__},
        custom include command={\includepdf{\robExtAddCachePathAndName{\robExtFinalHash.pdf}}}
    },
}

\begin{filecontents}{test2.tex}   
    \chapter{Test1}
    Test bla Test.   
\end{filecontents}

\begin{document}


\cacheMe[latex scrreport]{\input{__ROBEXT_WAY_BACK__test2.tex}}

\begin{CacheMe}{latex scrreport}
    \input{__ROBEXT_WAY_BACK__test2.tex}
\end{CacheMe}

\end{document}
```

**Reply by dflvunoooooo** [reply]
> But basically, when you create a preset from scratch, you always need to specify the compilation command, the template, and (often) the inclusion command. In your above code, you specified none of them. If you start it with `latex`, it will specify a template, a compilation command and an inclusion command, and to allow you to customize it, the template contains a number of placeholders, one of them containing the class. The above command therefore override the default standalone class to your own.

Thank you for that clarification. Makes sense that I would have to set up anything else, if I don't specify latex.

**Reply by tobiasBora** [reply]
Great. Just, if you use `\input`, make sure to add it to the dependencies, otherwise it will not be recompiled if you change the input file:
```
\begin{CacheMe}{latex scrreport, add dependencies=test2.tex}
    \input{__ROBEXT_WAY_BACK__test2.tex}
\end{CacheMe}
```

**Reply by tobiasBora** [reply]
Actualy, I just realized that I already created an easy command not to wrap the content, so you can use `do not wrap code` instead of the longer `set placeholder` line ^^

**Reply by tobiasBora** [reply]
Last remark: I'm not sure how you plan to include page numbering, but if you want the page numbers to update on change, you might want to disable page numbering in your cached file, and add them back via the appropriate options in `\includepdf` as discussed in https://tex.stackexchange.com/questions/21248/how-to-add-a-page-number-to-the-included-pdf-pages

**Reply by dflvunoooooo** [reply]
Thank you for your help! I did all this and it is working very well! I didn't think of the page numbers yet. A working example is
```
\documentclass{scrreport}


\usepackage{robust-externalize}
\usepackage{pdfpages}
\usepackage{blindtext}

\robExtConfigure{
    new preset={latex scrreport}{
        latex, 
        set documentclass={scrreport},
        set latex options={fontsize=11pt},
          add to preamble={\usepackage{scrlayer-scrpage} \clearpairofpagestyles},
        do not wrap code,
        custom include command={\includepdf[pagecommand={}]{\robExtAddCachePathAndName{\robExtFinalHash.pdf}}}
    },
}

\begin{filecontents}{test2.tex}   
    \chapter{Test1}
    Test bla Test. 
\end{filecontents}

\begin{document}

\Blindtext

\cacheMe[latex scrreport, add dependencies=test2.tex]{\input{__ROBEXT_WAY_BACK__test2.tex}}

\end{document}
```

**Reply by tobiasBora** [reply]
Btw, I'm thinking that before caching chapters, you might want to know a few caveats:
1. if you use internal links, you first need to use `set placeholder={__ROBEXT_LATEX_ENGINE__}{latexmk -pdflatex=lualatex -pdf},` instead of `use lualatex` in order to compile the document enough times for it to converge
2. sadly, this is not enough: `\includepdf` does not, out of the box, preserves links or annotations. So clicking on them will not move you anywhere. There is newpax that tries to avoid this issue https://github.com/u-fischer/newpax but I never spent much time to integrate it into robust-externalize. That is my next important improvement to this lib.
3. if you use links to other chapters… it is even worse, as it will think that the label is undefined. This might require non-trivial changes to this lib, I'm not sure here how to deal with this issue in a really clean way. I'll certainly get inspiration from `xr`, and need to parse the `.aux` file.
